src/middlewared/middlewared/plugins/enclosure_/m50_plx.py

```diff
@@ -1,58 +1,48 @@
+from pathlib import Path
 import re
 
+import pyudev
+
 from middlewared.service import Service, private
 
 
 class EnclosureService(Service):
-    RE_PCI = re.compile(r"pci([0-9]+)")
-    RE_PCIB = re.compile(r"pcib([0-9]+)")
-    RE_SLOT = re.compile(r"slot=([0-9]+)")
+    RE_SLOT = re.compile(r"^0-([0-9]+)$")
 
     @private
     def m50_plx_enclosures(self):
-        """
-        # TODO: fix on SCALE
         product = self.middleware.call_sync("system.dmidecode_info")["system-product-name"]
         if product is None or not ("TRUENAS-M50" in product or "TRUENAS-M60" in product):
             return []
 
-        nvme_to_nvd = self.middleware.call_sync('disk.nvme_to_nvd_map')
-
-        slot_to_nvd = {}
-        for nvme, nvd in nvme_to_nvd.items():
-            try:
-                pci = sysctl.filter(f"dev.nvme.{nvme}.%parent")[0].value
-                m = re.match(self.RE_PCI, pci)
-                if not m:
-                    continue
-
-                pcib = sysctl.filter(f"dev.pci.{m.group(1)}.%parent")[0].value
-                m = re.match(self.RE_PCIB, pcib)
-                if not m:
-                    continue
-
-                pnpinfo = sysctl.filter(f"dev.pcib.{m.group(1)}.%pnpinfo")[0].value
-                if "vendor=0x10b5 device=0x8717" not in pnpinfo:
-                    continue
-
-                location = sysctl.filter(f"dev.pcib.{m.group(1)}.%location")[0].value
-                m = re.match(self.RE_SLOT, location)
-                if not m:
-                    continue
-                slot = int(m.group(1))
-            except IndexError:
+        addresses_to_slots = {
+            (slot / "address").read_text().strip(): slot.name
+            for slot in Path("/sys/bus/pci/slots").iterdir()
+        }
+
+        slot_to_nvme = {}
+        context = pyudev.Context()
+        for nvme_device in context.list_devices(subsystem="nvme"):
```

Comment:
This for loop can be simplified quite a bit.
```
for nvme in filter(lambda x: x.parent.parent.get('PCI_SUBSYS_ID') == '10B5:8717', ctx.list_devices(subsystem='nvme')):
    if (slot := addrs_to_slots.get(nvme.parent.split('.')[0])) is not None:
        slot_to_nvme[int(slot.split('-', 1)[1])] = nvme.sys_name
```

However, if you choose to keep the logic in this loop then you're going to want to get the `DEVPATH` attribute via the `nvme_device.properties['DEVPATH']` because accessing it via `nvme_device` has been deprecated in py-udev and will be removed in py-udev version 1.0.0